pkg/R/utils.R

```diff
@@ -288,3 +254,117 @@ convertEnvsToList <- function(keys, vals) {
+               defVar <- node[[2]]
+               if (length(defVar) == 1 && typeof(defVar) == "symbol") {
+                 # Add the defined variable name into .defVars.
+                 assign(".defVars", 
+                        c(get(".defVars", envir = .sparkREnv), as.character(defVar)), 
+                        envir = .sparkREnv)
+               }
+               for (i in 3:nodeLen) {
+                 processClosure(node[[i]], oldEnv, argNames, newEnv)
+               }
+             },
+             "function" = {  # Function definition.
+               newArgs <- names(node[[2]])
+               argNames <- c(argNames, newArgs)  # Add parameter names.
+               for (i in 3:nodeLen) {
+                 processClosure(node[[i]], oldEnv, argNames, newEnv)
+               }
+             }, 
+             {
+               for (i in 1:nodeLen) {
+                 processClosure(node[[i]], oldEnv, argNames, newEnv)
+               }             
+            })
+    }
+  } else if (nodeLen == 1 && 
+               (typeof(node) == "symbol" || typeof(node) == "language")) {
+    # Base case: current AST node is a leaf node and a symbol or a function call.
+    nodeChar <- as.character(node)
+    if (!nodeChar %in% argNames && # Not a function parameter or function local variable.
+          !nodeChar %in% get(".defVars", envir = .sparkREnv)) {
+      func.env <- oldEnv
+      topEnv <- parent.env(.GlobalEnv)
+      # Search in function environment, and function's enclosing environments 
+      # up to global environment. There is no need to look into package environments
+      # above the global or namespace environment that is not SparkR below the global, 
+      # as they are assumed to be loaded on workers.
+      while (!identical(func.env, topEnv)) {
+        # Namespaces other than "SparkR" will not be searched.
+        if (!isNamespace(func.env) || getNamespaceName(func.env) == "SparkR") {
+          # Set parameter 'inherits' to FALSE since we do not need to search in
+          # attached package environments.
+          if (exists(nodeChar, envir = func.env, inherits = FALSE)) {
+            if (!isNamespace(func.env) || 
+                  !nodeChar %in% getNamespaceExports("SparkR")) {  # Only include SparkR internals.
+              obj <- get(nodeChar, envir = func.env)
+              if (is.function(obj)) {  
+                # if the node is a function call, recursively clean its closure.
+                obj <- cleanClosure(obj)
+              }
+              assign(nodeChar, obj, envir = newEnv)
+              break
+            }
+          }
+        }
+        
+        # Continue to search in enclosure.
+        func.env <- parent.env(func.env)
+      }
+    }
+  }
+}
+
+# Utility function to get user defined function (UDF) dependencies (closure). 
+# More specifically, this function captures the values of free variables defined 
+# outside a UDF, and stores them in a new environment.
```

Comment:
In the last line `stores them in the function's environment` might be more accurate